pkgs/extension_discovery/lib/extension_discovery.dart

```diff
@@ -0,0 +1,244 @@
+
+  Extension._({
+    required this.package,
+    required this.rootUri,
+    required this.packageUri,
+    required this.config,
+  });
+}
+
+/// Find extensions for [targetPackage] provided by packages in
+/// `.dart_tool/package_config.json`.
+///
+/// ## Locating `.dart_tool/package_config.json`
+///
+/// This method requires the location of the [packageConfig], unless the current
+/// isolate has been setup for package resolution.
+/// Notably, Dart programs compiled for AOT cannot find their own
+/// `package_config.json`.
+///
+/// If operating on a project that isn't the current project, for example, if
+/// you are developing a tool that users are globally activating and then
+/// running against their own projects, and you wish to detect extensions within
+/// their projects, then you must specify the path the
+/// `.dart_tool/package_config.json` for the users project as [packageConfig].
+///
+/// ## Detection of extensions
+///
+/// An extension for [targetPackage] is detected in `package:foo` if `foo`
+/// contains `extension/<targetPackage>/config.json`, and the contents of this
+/// file is valid JSON.
+///
+/// ### Caching results
+///
+/// When [useCache] is `true` then the detected extensions will be cached
+/// in `.dart_tool/extension_discovery/<targetPackage>.json`.
+/// This function will compare modification timestamps of
+/// `.dart_tool/package_config.json` with the cache file, before reusing cached
+/// results.
+/// This function will also treat relative path-dependencies as mutable
+/// packages, and check such packages for extensions every time [findExtensions]
+/// is called. Notably, it'll compare the modification time of the
+/// `extension/<targetPackage>/config.json` file, to ensure that it's older than
+/// the extension cache file. Otherwise, it'll reload the extension
+/// configuration.
+///
+/// ## Exceptions
+///
+/// This method will throw [PackageConfigException], if the
+/// `.dart_tool/package_config.json` file specified in [packageConfig] could not
+/// be loaded or is invalid. This usually happens if dependencies are not
+/// resolved, and users can probably address it by running `dart pub get`.
+///
+/// But, **do consider catch** [PackageConfigException] and handling the failure
+/// to load extensions appropriately.
+///
+/// This method will throw an [Error] if [packageConfig] is not specified, and
+/// the current isolate isn't configured for package resolution.
+Future<List<Extension>> findExtensions(
+  String targetPackage, {
+  bool useCache = true,
+  Uri? packageConfig,
+}) async {
+  final packageConfigUri = packageConfig ?? await Isolate.packageConfig;
+  if (packageConfigUri == null) {
+    throw UnsupportedError(
+      'packageConfigUri must be provided, if not running in JIT mode',
+    );
+  }
+  final packageConfigFile = File.fromUri(packageConfigUri);
+  final registryFile = File.fromUri(packageConfigFile.parent.uri.resolve(
+    'extension_discovery/$targetPackage.json',
+  ));
```

Comment:
just to clarify understanding, is this is what a directory structure for `some_app` could end up looking like after calling `findExtensions`?
```
findExtensions(
  'devtools',
  packageConfig: 'path/to/some_app/.dart_tool/package_config.json',
);
```
resulting file system structure for `some_app`:
```
some_app/
    lib/
        .dart_tool/
            package_config.json
            extension_discovery/
                foo.json
                bar.json
                provider.json
                riverpod.json
```

Reply:
`.dart_tool/` is on the same level as `lib/`, `pubspec.yaml`, etc...

```
some_app/
    pubspec.yaml
    pubspec.lock
    bin/
    test/
    lib/
    .dart_tool/
        package_config.json
        extension_discovery/
            devtools.json
            analyzer.json
            some_extendable_package.json
            some_other_extendable_package.json
```

But yes. Noting that files in `extension_discovery/` are generated on-demand.